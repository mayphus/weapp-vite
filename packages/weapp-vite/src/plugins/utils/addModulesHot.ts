import type { PluginContext } from 'rollup'

export function addModulesHot(entriesSet: Set<string>, pluginContext: PluginContext) {
  for (const entry of entriesSet) {
    const moduleInfo = pluginContext.getModuleInfo(entry)

    if (moduleInfo) {
      const stack = [moduleInfo.id] // 用栈模拟递归
      const visitedModules = new Set<string>()

      while (stack.length > 0) {
        const id = stack.pop()

        if (id && !visitedModules.has(id)) {
          visitedModules.add(id)

          const info = pluginContext.getModuleInfo(id)

          if (info) {
            pluginContext.addWatchFile(info.id)
            // 将子依赖加入栈
            stack.push(...info.importedIds)
          }
        }
      }
    }
  }
}
