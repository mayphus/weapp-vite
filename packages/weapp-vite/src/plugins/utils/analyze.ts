import type { App, Component, Page } from '@weapp-core/schematics'
import { get } from '@weapp-core/shared'

export function analyzeAppJson(json: App) {
  const entries: string[] = []
  const pages = json.pages ?? []
  const components = Object.values(json.usingComponents ?? {})
  const subPackages = (
    [...json.subPackages ?? [], ...json.subpackages ?? []]?.reduce<string[]>(
      (acc, cur) => {
        acc.push(...(cur.pages ?? []).map(x => `${cur.root}/${x}`))
        return acc
      },
      [],
    )) ?? []

  if (get(json, 'tabBar.custom')) {
    entries.push('custom-tab-bar/index')
  }
  // 全局工具栏
  // https://developers.weixin.qq.com/miniprogram/dev/framework/runtime/skyline/appbar.html
  if (get(json, 'appBar')) {
    entries.push('app-bar/index')
  }
  entries.push(
    ...pages,
    ...components,
    ...subPackages,
  )
  return entries
}

export function analyzeCommonJson(json: Page | Component) {
  const entries: string[] = []
  const components = Object.values(json.usingComponents ?? {})
  entries.push(...components)
  return entries
}
