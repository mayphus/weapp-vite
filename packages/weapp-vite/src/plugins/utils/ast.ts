import type { Node } from 'estree'
// import type { ProgramNode } from 'rollup'
import { walk } from 'estree-walker'

export function collectRequireTokens(ast: Node) {
  const requireModules: {
    start: number
    end: number
    value: string
    async?: boolean
    // leadingComment: string
  }[] = []
  walk(ast, {
    enter(node) {
      if (node.type === 'CallExpression') {
        // TODO 可能要被废弃
        // https://developers.weixin.qq.com/miniprogram/dev/reference/api/require.html#require-async-%E9%93%BE%E5%BC%8F%E8%B0%83%E7%94%A8
        if (node.callee.type === 'Identifier' && node.callee.name === 'require') {
          if (node.arguments[0] && node.arguments[0].type === 'Literal' && typeof node.arguments[0].value === 'string') {
            requireModules.push({
              // @ts-ignore
              start: node.arguments[0].start,
              // @ts-ignore
              end: node.arguments[0].end,
              value: node.arguments[0].value,
            })
          }
        }
        else if (
          node.callee.type === 'MemberExpression'
          && node.callee.object.type === 'Identifier'
          && node.callee.object.name === 'require'
          && node.callee.property.type === 'Identifier'
          && node.callee.property.name === 'async') {
          if (node.arguments[0] && node.arguments[0].type === 'Literal' && typeof node.arguments[0].value === 'string') {
            requireModules.push({
              // @ts-ignore
              start: node.arguments[0].start,
              // @ts-ignore
              end: node.arguments[0].end,
              value: node.arguments[0].value,
              async: true,
            })
          }
        }
      }
    },
  })

  return {
    requireModules,
  }
}
