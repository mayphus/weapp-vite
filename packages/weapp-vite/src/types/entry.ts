import type { App as AppJson, Component as ComponentJson, Page as PageJson, Sitemap, Theme } from '@weapp-core/schematics'

export interface BaseEntry {
  path: string
  jsonPath?: string
  json?: object
  type: 'app' | 'page' | 'component' | (string & {})
}

export type Entry = AppEntry | PageEntry | ComponentEntry

/**
 * 应用入口 js + json
 */
export interface AppEntry extends BaseEntry {
  type: 'app'
  themeJsonPath?: string
  themeJson?: Theme
  sitemapJsonPath?: string
  sitemapJson?: Sitemap
  json: AppJson
  jsonPath: string
}

/**
 * 页面入口 js + wxml
 */
export interface PageEntry extends BaseEntry {
  type: 'page'
  templatePath: string
  json?: PageJson
}

/**
 * 组件入口 js + wxml + json + json.component === true
 */
export interface ComponentEntry extends BaseEntry {
  type: 'component'
  templatePath: string
  json: ComponentJson
  jsonPath: string
}

export type EntryJsonFragment = Omit<BaseEntry, 'path' | 'type'>
