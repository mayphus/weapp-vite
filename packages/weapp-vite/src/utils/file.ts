import { configExtensions, jsExtensions, templateExtensions, vueExtensions } from '@/constants'
import fs from 'fs-extra'
import path from 'pathe'

export function isJsOrTs(name?: string) {
  if (typeof name === 'string') {
    return jsExtensions.some(x => name.endsWith(`.${x}`))
  }
  return false
}

export function isTemplateRequest(request: string) {
  return request.endsWith('.wxml') || request.endsWith('.html')
}

export function changeFileExtension(filePath: string, extension: string) {
  if (typeof filePath !== 'string') {
    throw new TypeError(`Expected \`filePath\` to be a string, got \`${typeof filePath}\`.`)
  }

  if (typeof extension !== 'string') {
    throw new TypeError(`Expected \`extension\` to be a string, got \`${typeof extension}\`.`)
  }

  if (filePath === '') {
    return ''
  }

  extension = extension ? (extension.startsWith('.') ? extension : `.${extension}`) : ''

  const basename = path.basename(filePath, path.extname(filePath))
  return path.join(path.dirname(filePath), basename + extension)
}

export async function findVueEntry(filepath: string) {
  for (const ext of vueExtensions) {
    const p = changeFileExtension(filepath, ext)
    if (await fs.exists(p)) {
      return p
    }
  }
}

export async function findJsEntry(filepath: string) {
  for (const ext of jsExtensions) {
    const p = changeFileExtension(filepath, ext)
    if (await fs.exists(p)) {
      return p
    }
  }
}

export async function findJsonEntry(filepath: string) {
  for (const ext of configExtensions) {
    const p = changeFileExtension(filepath, ext)
    if (await fs.exists(p)) {
      return p
    }
  }
}

export async function findTemplateEntry(filepath: string) {
  for (const ext of templateExtensions) {
    const p = changeFileExtension(filepath, ext)
    if (await fs.exists(p)) {
      return p
    }
  }
}

export function isTemplate(filepath: string) {
  return templateExtensions.some(ext => filepath.endsWith(`.${ext}`))
}

export function touch(filename: string) {
  const time = new Date()

  try {
    fs.utimesSync(filename, time, time)
  }
  catch {
    fs.closeSync(fs.openSync(filename, 'w'))
  }
}
