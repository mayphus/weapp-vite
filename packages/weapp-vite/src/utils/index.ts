export * from './file'
export * from './json'
export * from './projectConfig'
export * from './scan'
// export * from './wxml'
// export * from './wxs'
