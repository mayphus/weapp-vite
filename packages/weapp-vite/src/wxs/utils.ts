import { addExtension, removeExtensionDeep } from '@weapp-core/shared'

export function normalizeWxsFilename(value: string) {
  return addExtension(removeExtensionDeep(value), '.wxs')
}
