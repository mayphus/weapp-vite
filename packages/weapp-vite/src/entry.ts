import type { Entry, WeappViteConfig } from './types'
import process from 'node:process'
import { defu } from '@weapp-core/shared'
import klaw from 'klaw'
import mm from 'micromatch'
import path from 'pathe'
import { defaultExcluded } from './defaults'
import { getWxmlEntry, searchAppEntry } from './utils'

export function createFilter(include: string[], exclude: string[], options?: mm.Options) {
  const opts = defu<mm.Options, mm.Options[]>(options, {
    ignore: exclude,
  })

  return function (id: unknown | string) {
    if (typeof id !== 'string') {
      return false
    }
    if (/\0/.test(id)) {
      return false
    }

    return mm.isMatch(id as string, include, opts)
  }
}

export async function getEntries(options: { root?: string, srcRoot?: string, outDir?: string, relative?: boolean, subPackage?: WeappViteConfig['subPackage'] }) {
  // build.outDir
  const { root = process.cwd(), outDir = 'dist', relative, srcRoot = '', subPackage } = options

  function formatPath(to: string) {
    if (relative) {
      return path.relative(root, to)
    }
    return path.normalize(to)
  }

  // 单独打包分包的场景
  if (subPackage) {
    const subPackageRoot = subPackage.root ?? ''
    const filter = createFilter(
      [path.join(root, srcRoot, subPackageRoot, '**/*')],
      [
        ...defaultExcluded,
      ],
      { cwd: root },
    )

    const pageEntries: Entry[] = []
    const componentEntries: Entry[] = []
    const subPackageEntries: Entry[] = []

    if (subPackage.entry) {
      subPackageEntries.push({
        deps: [],
        path: path.join(root, subPackageRoot, subPackage.entry),
        type: 'subPackageEntry',
      })
    }

    for await (
      const file of klaw(
        path.join(root, subPackageRoot),
        {
          filter,
        },
      )
    ) {
      if (file.stats.isFile()) {
        if (/\.wxml$/.test(file.path)) {
          const entry = getWxmlEntry(file.path, formatPath)
          if (entry) {
            //  && !walkPathsSet.has(file.path)
            // 防止重复，理论上不会
            // walkPathsSet.add(file.path)
            if (entry.type === 'component') {
              componentEntries.push(entry)
            }
            else if (entry.type === 'page') {
              pageEntries.push(entry)
            }
          }
        }
      }
    }
    return {
      pages: pageEntries,
      components: componentEntries,
      subPackageEntries,
    }
  }
  // 打包主包的场景
  else {
    const appEntry = searchAppEntry({
      root: path.join(root, srcRoot),
      formatPath,
    })

    // TODO exclude 需要和 output 配套
    // const walkPathsSet = new Set<string>()
    if (appEntry) {
      const subPackageDeps = appEntry.deps.filter(x => x.type === 'subPackage')
      const filter = createFilter(
        [path.join(root, srcRoot, '**/*')],
        [
          ...defaultExcluded,
          path.join(root, `${outDir}/**`),
          ...subPackageDeps.map((x) => {
            return path.join(root, `${x.root}/**`)
          }),
        ],
        { cwd: root },
      )

      const pageEntries: Entry[] = []
      const componentEntries: Entry[] = []

      for await (
        const file of klaw(
          path.join(root, srcRoot),
          {
            filter,
          },
        )
      ) {
        if (file.stats.isFile()) {
          if (/\.wxml$/.test(file.path)) {
            const entry = getWxmlEntry(file.path, formatPath)
            if (entry) {
              //  && !walkPathsSet.has(file.path)
              // 防止重复，理论上不会
              // walkPathsSet.add(file.path)
              if (entry.type === 'component') {
                componentEntries.push(entry)
              }
              else if (entry.type === 'page') {
                pageEntries.push(entry)
              }
            }
          }
        }
      }

      return {
        app: appEntry,
        pages: pageEntries,
        components: componentEntries,
        subPackages: subPackageDeps,
        // walkPathsSet,
      }
    }
  }
}
