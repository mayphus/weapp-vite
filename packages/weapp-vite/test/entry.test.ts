import { getEntries } from '@/entry'
import { absDirs } from './utils'

describe('entry', () => {
  // function normalizeScanEntries(result: Awaited<ReturnType<typeof scanEntries>>) {
  //   return {
  //     app: path.normalize(result?.app ?? ''),
  //     pages: [...result?.pages ?? []].map(x => path.normalize(x)),
  //     components: [...result?.components ?? []].map(x => path.normalize(x)),
  //     // css: result?.css.map(x => path.normalize(x)),
  //   }
  // }
  describe('scanEntries', () => {
    it.each(absDirs)('$name', async ({ path: root }) => {
      // expect(normalizeScanEntries()).toMatchSnapshot()
      const entries = await getEntries({
        root,
        relative: true,
      })
      expect(entries).toMatchSnapshot()
    })
  })
})
