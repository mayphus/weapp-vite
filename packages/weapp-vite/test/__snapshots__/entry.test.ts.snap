// Vitest Snapshot v1, https://vitest.dev/guide/snapshot.html

exports[`entry > scanEntries > 'vite-native' 1`] = `
{
  "app": {
    "deps": [
      {
        "path": "pages/index/index",
        "type": "page",
      },
      {
        "path": "/components/navigation-bar/navigation-bar",
        "type": "component",
      },
      {
        "entry": "index.js",
        "name": "pack1",
        "pages": [
          "pages/cat",
          "pages/dog",
        ],
        "root": "packageA",
        "type": "subPackage",
      },
      {
        "entry": "index.ts",
        "independent": true,
        "name": "pack2",
        "pages": [
          "pages/apple",
          "pages/banana",
        ],
        "root": "packageB",
        "type": "subPackage",
      },
    ],
    "json": {
      "componentFramework": "glass-easel",
      "lazyCodeLoading": "requiredComponents",
      "pages": [
        "pages/index/index",
      ],
      "sitemapLocation": "sitemap.json",
      "subPackages": [
        {
          "entry": "index.js",
          "name": "pack1",
          "pages": [
            "pages/cat",
            "pages/dog",
          ],
          "root": "packageA",
        },
        {
          "entry": "index.ts",
          "independent": true,
          "name": "pack2",
          "pages": [
            "pages/apple",
            "pages/banana",
          ],
          "root": "packageB",
        },
      ],
      "usingComponents": {
        "navigation-bar": "/components/navigation-bar/navigation-bar",
      },
      "window": {
        "navigationBarBackgroundColor": "#ffffff",
        "navigationBarTextStyle": "black",
        "navigationBarTitleText": "Weixin",
      },
    },
    "jsonPath": "app.json",
    "path": "app.js",
    "type": "app",
  },
  "components": [
    {
      "deps": [],
      "json": {
        "component": true,
        "styleIsolation": "apply-shared",
        "usingComponents": {},
      },
      "jsonPath": "components/navigation-bar/navigation-bar.json",
      "path": "components/navigation-bar/navigation-bar.ts",
      "type": "component",
    },
  ],
  "pages": [
    {
      "deps": [
        {
          "path": "tdesign-miniprogram/button/button",
          "type": "component",
        },
        {
          "path": "tdesign-miniprogram/divider/divider",
          "type": "component",
        },
      ],
      "json": {
        "usingComponents": {
          "t-button": "tdesign-miniprogram/button/button",
          "t-divider": "tdesign-miniprogram/divider/divider",
        },
      },
      "jsonPath": "pages/index/index.json",
      "path": "pages/index/index.ts",
      "type": "page",
    },
  ],
  "subPackages": [
    {
      "entry": "index.js",
      "name": "pack1",
      "pages": [
        "pages/cat",
        "pages/dog",
      ],
      "root": "packageA",
      "type": "subPackage",
    },
    {
      "entry": "index.ts",
      "independent": true,
      "name": "pack2",
      "pages": [
        "pages/apple",
        "pages/banana",
      ],
      "root": "packageB",
      "type": "subPackage",
    },
  ],
}
`;

exports[`entry > scanEntries > 'vite-native-skyline' 1`] = `
{
  "app": {
    "deps": [
      {
        "path": "pages/index/index",
        "type": "page",
      },
    ],
    "json": {
      "componentFramework": "glass-easel",
      "lazyCodeLoading": "requiredComponents",
      "pages": [
        "pages/index/index",
      ],
      "renderer": "skyline",
      "rendererOptions": {
        "skyline": {
          "defaultDisplayBlock": true,
          "disableABTest": true,
          "sdkVersionBegin": "3.0.0",
          "sdkVersionEnd": "15.255.255",
        },
      },
      "sitemapLocation": "sitemap.json",
      "style": "v2",
      "window": {
        "navigationBarTextStyle": "black",
        "navigationStyle": "custom",
      },
    },
    "jsonPath": "app.json",
    "path": "app.js",
    "type": "app",
  },
  "components": [
    {
      "deps": [],
      "json": {
        "component": true,
        "styleIsolation": "apply-shared",
        "usingComponents": {},
      },
      "jsonPath": "components/navigation-bar/navigation-bar.json",
      "path": "components/navigation-bar/navigation-bar.js",
      "type": "component",
    },
  ],
  "pages": [
    {
      "deps": [
        {
          "path": "/components/navigation-bar/navigation-bar",
          "type": "component",
        },
      ],
      "json": {
        "usingComponents": {
          "navigation-bar": "/components/navigation-bar/navigation-bar",
        },
      },
      "jsonPath": "pages/index/index.json",
      "path": "pages/index/index.js",
      "type": "page",
    },
  ],
  "subPackages": [],
}
`;

exports[`entry > scanEntries > 'vite-native-ts' 1`] = `undefined`;

exports[`entry > scanEntries > 'vite-native-ts-skyline' 1`] = `undefined`;
